openevsewifi/__init__.py

```diff
@@ -27,22 +28,65 @@
 
 colors = ['off', 'red', 'green', 'yellow', 'blue', 'violet', 'teal', 'white']
 
+def parse_checksum(s):
+    """
+    If there is a '^' in given string s, this checks that the xor of utf8 bytes
+    before the '^' equal the hex value specified after '^'.  It returns the
+    string before the '^' on success, None on error.
+
+    If there is no '^' in the string, the string is returned.
+    """
+    spl = s.rsplit('^', 1)
+    if len(spl) == 1:
+        return s
+    try:
+        check = int(spl[1], 16)
+    except:
+        return None
+    datsum = 0
+    for c in spl[0].encode('utf-8'):
+        datsum ^= c
+    if datsum != check:
+        return None
+    return spl[0]
+
+def json_parser(s):
+    print("json: ", s)
+    try:
+        result = json.loads(s)
+        print("result: ", result)
+        parsed = parse_checksum(result["ret"])
+        if parsed is not None:
+            parsed = parsed[1:] # strip of leading $
+            parsed = parsed.split()
+        return parsed
+    except Exception as e:
+        print(e)
+        return None
+
+def xml_parser(s):
+    response = re.search('\\<p>&gt;\\$([^\\^]+)(\\^..)?<script', s)
+    # If we are using version 1
+    # https://github.com/OpenEVSE/ESP8266_WiFi_v1.x/blob/master/OpenEVSE_RAPI_WiFi_ESP8266.ino#L357
+    if response is None:
+        response = re.search('\\>\\>\\$(.+)\\<p>', s)
+    return response.group(1).split()
 
 class Charger:
-    def __init__(self, host: str):
+    def __init__(self, host: str, json=False):
         """A connection to an OpenEVSE charging station equipped with the wifi kit."""
-        self._url = 'http://' + host + '/r?'
+        if json:
+            self._url = 'http://' + host + '/r?json=1&'
+            self._parseResult = json_parser
+        else:
+            self._url = 'http://' + host + '/r?'
+            self._parseResult = xml_parser
 
     def _send_command(self, command: str) -> List[str]:
         """Sends a command through the web interface of the charger and parses the response"""
         data = {'rapi': command}
         content = requests.post(self._url, data=data)
-        response = re.search('\\<p>&gt;\\$([^\\^]+)(\\^..)?<script', content.text)
-        # If we are using version 1
-        # https://github.com/OpenEVSE/ESP8266_WiFi_v1.x/blob/master/OpenEVSE_RAPI_WiFi_ESP8266.ino#L357
-        if response is None:
-            response = re.search('\\>\\>\\$(.+)\\<p>', content.text)
-        return response.group(1).split()
+        return self._parseResult(content.text)
```

Comment:
The type hint on line 85 indicates that `_send_command` should return a list of strings, but your change is violating that contract - it now returns a string if the parser is XML, but if it is using the JSON parser and runs into an error, it could return None.